pkg/trace/obfuscate/sql.go

```diff
@@ -342,3 +344,73 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
 	}
 	traceutil.SetMeta(span, sqlQueryTag, oq.Query)
 }
+
+type mapEntryTransform func(key string, value interface{}) (string, interface{})
+
+// transformMapEntriesRecursive returns a copy of the map object, recursively transforming any keys matching keyPattern
+func transformMapEntriesRecursive(obj interface{}, keyPattern *regexp.Regexp, transform mapEntryTransform) interface{} {
+	if mapObj, ok := obj.(map[string]interface{}); ok {
+		result := make(map[string]interface{})
+		for key, val := range mapObj {
+			if keyPattern.MatchString(key) {
+				key, val = transform(key, val)
+				// transform can decide to drop a key by returning empty string for key
+				if key != "" {
+					result[key] = val
+				}
+			} else {
+				result[key] = transformMapEntriesRecursive(val, keyPattern, transform)
+			}
+		}
+		return result
+	} else if listObj, ok := obj.([]interface{}); ok {
+		result := make([]interface{}, len(listObj), len(listObj))
+		for i, val := range listObj {
+			result[i] = transformMapEntriesRecursive(val, keyPattern, transform)
+		}
+		return result
+	}
+	return obj
+}
+
+var planObfuscationPattern = regexp.MustCompile(`attached_condition|(\w+ Cond)`)
+
+// ObfuscateSQLPlan obfuscates all query conditions (i.e. (where id=5) -> (where id=?)) in the provided json execution
+// plan.
+func (o *Obfuscator) ObfuscateSQLPlan(plan map[string]interface{}) map[string]interface{} {
+	result := transformMapEntriesRecursive(plan, planObfuscationPattern, func(key string, value interface{}) (string, interface{}) {
+		strValue, ok := value.(string)
+		if !ok {
+			log.Errorf("wrong type found for key '%s'. Expected string, Found: %v", key, reflect.TypeOf(value))
+			return key, "?"
+		}
+		obfQuery, err := o.ObfuscateSQLString(strValue)
+		if err != nil {
+			log.Errorf("failed to obfuscate sql plan: %s", err.Error())
```

Comment:
I think these logs could be noisy and the returned values result in false positives of a plan "changing" even when it was just an error. What do you think about instead returning an error for the whole plan for these cases? Then we would throw it away rather than submit a plan with overly-obfuscated values.

Reply:
> I think these logs could be noisy

Agreed, updated to send debug logs instead. 

> Then we would throw it away rather than submit a plan with overly-obfuscated values.

Why is it better to throw it away than to overly obfuscate? Obfuscation failures really shouldn't happen because that would mean we're probably failing to obfuscate the full query samples that contain that condition as well. 

What do you think of putting an error message in the obfuscated value, that way the plan still gets ingested enabling us to detect that the failed obfuscation is happening? 

See here https://github.com/DataDog/datadog-agent/pull/5692/commits/40a54b3cb2bcee2b5d1bfb96366ec70c9eec4ef0

Reply:
> Obfuscation failures really shouldn't happen because that would mean we're probably failing to obfuscate the full query samples that contain that condition as well.

Agreed, it would mean we missed a case, but it can and will happen once we expose it to users' queries. We need a way to discover these since this will be behind customer infrastructure. Maybe we should emit a metric.

> What do you think of putting an error message in the obfuscated value, that way the plan still gets ingested enabling us to detect that the failed obfuscation is happening?

Since that plan will go to users, I don't think we should do that. Ultimately it's an error on our part, not something they can fix. So the workflow would be for them to see a plan with that error message, enable debug logs on the agent, and then tell us what the error message says? I'm not a fan of that solution.

If we choose to show the plan (sending '?' on the key that error'ed), we need a way to signal that it's semi-correct. I would rather see the errors surfaced up to the outer function and attach it as another field to the log message so we could show a ⚠️  or something in the UI. Then it's best-effort to display and we can track error frequency.